fix(core): Enhance loop detection and prevent infinite execution in workflows

cstuncsik opened the pull request:
## Summary

Executing a Loop Over Items node connecting its output to its own input crashes n8n instance.

<img width="1472" height="852" alt="image" src="https://github.com/user-attachments/assets/8113be7c-4b57-4fe4-abde-1e13b29ea439" />

## Related Linear tickets, Github issues, and Community forum posts

PAY-2940

## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds structural and runtime guards to stop SplitInBatches self-loop infinite executions, introduces getOutgoingConnections/getIncomingConnections, and tightens engine-side loop detection.
> 
> - **Core/Engine**:
>   - Structural guard in `workflow-execute.ts` to block definite infinite loops (SplitInBatches "done" -> self input); converts immediate-loop error to `OperationalError`.
>   - Adds `SPLIT_IN_BATCHES_NODE_TYPE` constant in `workflow/constants.ts`.
> - **Execution Context**:
>   - New graph introspection APIs: `getOutgoingConnections()` and `getIncomingConnections()` in `node-execution-context.ts` with tests.
> - **SplitInBatches Node**:
>   - Registers v4 (`v4/SplitInBatchesV4.node.ts`) with runtime execution counters, structural self-loop guard, reset/cleanup utilities, and configurable options (disable protection, limitFactor, maxReinitializations).
>   - Adds structural guard to v3 implementation.
>   - Extensive unit tests for v3/v4 and infinite-loop protection.
> - **Types/Interfaces**:
>   - Extends `FunctionsBase` with graph introspection methods; adds `IWorkflowValidationWarning` and `validateWorkflowStructure` setting.
> - **Tests/Minor**:
>   - New tests for context graph APIs; small jest cast fix in CLI manual execution test.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 7a07f2a74b789c80902de5180023961b58b83036. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

cstuncsik committed:
fix(core): Enhance loop detection and prevent infinite execution in workflows

cstuncsik committed:
fix(test): Update type casting for WorkflowExecute mock implementation

cstuncsik committed:
fix(core): Remove loop detection configuration and enhance execution limits in SplitInBatches node

cstuncsik committed:
fix(core): Remove unused imports from workflow execution files

cstuncsik committed:
test: Remove enhanced loop detection tests to streamline workflow execution tests

cstuncsik committed:
fix(core): Implement infinite loop protection in SplitInBatches node execution

cstuncsik committed:
fix(test): Update type casting for mockExecuteFunctions in SplitInBatches node tests

cstuncsik committed:
fix(test): Update string interpolation to use single quotes in SplitInBatches node tests

cstuncsik committed:
Merge remote-tracking branch 'origin/master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
Merge remote-tracking branch 'origin/master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
fix: Enhance infinite loop protection with dynamic execution limits and ensure counter reset after normal completion

cstuncsik committed:
fix: Improve infinite loop handling by resetting execution counters and adding cleanup method

cstuncsik committed:
fix: Change cleanupExecutionCounters method to static for better memory management

cstuncsik commented:
@claude

cstuncsik committed:
fix: Implement automatic cleanup for abandoned executions to prevent memory leaks

cstuncsik committed:
fix: Add periodic cleanup timer for abandoned executions and allow timer to be cleared on shutdown

elsmr commented on packages/nodes-base/nodes/SplitInBatches/v3/SplitInBatchesV3.node.ts:
```diff
@@ -5,9 +5,39 @@ import type {
+				const timeout = 5 * 60 * 1000; // 5 minutes
+
+				for (const [executionId, timestamp] of SplitInBatchesV3.executionTracking) {
+					if (now - timestamp > timeout) {
```
Not sure if this will be reliable, workflows could take longer than 5mins between loops, especially with Wait nodes or HITL.

elsmr commented on packages/nodes-base/nodes/SplitInBatches/v3/SplitInBatchesV3.node.ts:
```diff
@@ -154,11 +197,81 @@ export class SplitInBatchesV3 implements INodeType {
+		// Minimum of 10 to handle small datasets, maximum of 1000 to prevent true infinite loops
+		const expectedBatches = Math.ceil(inputItems.length / Math.max(batchSize, 1));
+		const bufferMultiplier = 2; // Allow 2x expected batches for safety
+		const maxExecutions = Math.min(Math.max(expectedBatches * bufferMultiplier, 10), 1000);
```
Not sure if this will hold up in all cases, I could see cases where the done output is connected back to the input of the loop with an If node in between, and it could generate more executions than `maxExecutions` without being an infinite loop.

elsmr commented on packages/nodes-base/nodes/SplitInBatches/v3/SplitInBatchesV3.node.ts:
```diff
@@ -154,11 +197,81 @@ export class SplitInBatchesV3 implements INodeType {
+		const bufferMultiplier = 2; // Allow 2x expected batches for safety
+		const maxExecutions = Math.min(Math.max(expectedBatches * bufferMultiplier, 10), 1000);
+
+		if (newCount > maxExecutions) {
```
I am worried that if the `maxExecutions` is wrong, like in the case in my previous comment, we break existing workflows. We should at least version this and maybe add a way to disable it.

elsmr reviewed (commented):
This is great work but it's a [hard problem](https://en.wikipedia.org/wiki/Halting_problem) to solve:  and I don't think it's possible to make it 100% accurate. We should at least version the node, and provide an escape hatch in parameter options.

Another approach could be to detect the done output being connected back to the input of the loop node. If it's connected directly it's definitely an infinite loop, otherwise it gets more complicated.

cstuncsik committed:
Merge remote-tracking branch 'origin/master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
feat(editor): Introduce SplitInBatchesV4 node with infinite loop protection enhancements

cstuncsik committed:
Merge remote-tracking branch 'origin/master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
Merge remote-tracking branch 'origin/master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
fix: Prevent infinite loops by adding checks for self-referencing connections and enhance execution limit handling

cstuncsik committed:
fix: Reorganize type imports for better readability and maintainability

cstuncsik committed:
fix: Reset all counters before throwing error to avoid stale state

elsmr reviewed (commented):
Looking good! just 2 small comments and some conflicts

cstuncsik committed:
fix: Remove unused IWorkflowValidationResult interface to clean up code

cstuncsik committed:
Merge remote-tracking branch 'origin/master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
fix: Refactor imports in workflow-execute.ts for better organization

cstuncsik committed:
fix: Reorganize imports in workflow-execute.ts for improved clarity

cstuncsik committed:
feat: Implement graph introspection API for outgoing and incoming connections in NodeExecutionContext

cstuncsik committed:
fix: Import NodeConnectionTypes for consistency in node execution context

cstuncsik committed:
fix: Adjust infinite loop protection to check before resetting counters in SplitInBatchesV4

cstuncsik committed:
fix: Clean up type assertions and reorganize imports in node-execution-context

cstuncsik committed:
fix: Update infinite loop protection logic to allow up to 2x execution limit and clean up related parameters

cstuncsik committed:
feat: Add option to disable infinite loop protection and implement cleanup for SplitInBatches counters

cstuncsik committed:
fix: Replace OperationalError with UserError for infinite loop detection in SplitInBatches node
fix: Update constants to use SplitInBatches instead of SplitInBatchesV4 for consistency
fix: Adjust SplitInBatchesV4 completion logic to reset execution count after preparing result

cstuncsik committed:
fix: Replace OperationalError with ApplicationError for endless loop detection and enhance infinite loop protection handling in SplitInBatchesV4

cstuncsik committed:
fix: Enhance infinite loop detection logic in SplitInBatchesV4 by adjusting execution limit based on stuck detection

cstuncsik committed:
Merge remote-tracking branch 'origin/master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
fix: Add allowDisableLoopProtection option to GlobalConfig

cstuncsik committed:
fix: Update type imports and connections in infinite loop tests for consistency

cstuncsik committed:
fix: Update SplitInBatches default version to 4 and ensure proper cleanup of execution counters

cstuncsik committed:
fix: Enhance error logging for SplitInBatches cleanup and update test comments for clarity

cstuncsik committed:
Merge remote-tracking branch 'origin/master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
refactor: Simplify cleanup logic for SplitInBatches node type in WorkflowExecute

cstuncsik committed:
fix: Update SplitInBatches node type references and enhance loop protection description

cstuncsik committed:
fix: Enhance SplitInBatches node's infinite loop protection and improve limit factor handling

cstuncsik committed:
fix: Improve expected batches calculation in SplitInBatchesV4 to utilize maxRunIndex from context

cstuncsik committed:
Merge branch 'master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
Merge branch 'master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
Merge branch 'master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
Merge branch 'master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
Merge branch 'master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
Merge remote-tracking branch 'origin/master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
Merge remote-tracking branch 'origin/master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n

cstuncsik committed:
Merge remote-tracking branch 'origin/master' into pay-2940-bug-loop-node-connecting-its-output-to-input-crashes-n8n
